include/aos/iam/certmodule.hpp

```diff
@@ -0,0 +1,380 @@
+
+    /**
+     * Applies certificate chain to a module.
+     *
+     * @param certChain certificate chain.
+     * @param[out] topCertInfo info about a top certificate in a chain.
+     * @param[out] password owner password.
+     * @return Error.
+     */
+    virtual Error ApplyCertChain(
+        const Array<crypto::x509::Certificate>& certChain, CertInfo& topCertInfo, String& password)
+        = 0;
+
+    /**
+     * Removes certificate chain using top level certificate URL and password.
+     *
+     * @param certURL top level certificate URL.
+     * @param password owner password.
+     * @return Error.
+     */
+    virtual Error RemoveCertChain(const String& certURL, const String& password) = 0;
+
+    /**
+     * Removes private key from a module.
+     *
+     * @param keyURL private key URL.
+     * @param password owner password.
+     * @return Error.
+     */
+
+    virtual Error RemoveKey(const String& keyURL, const String& password) = 0;
+
+    /**
+     * Validates certificates in a module and returns information about valid/invalid certificates.
+     *
+     * @param[out] validInfo information about valid certificates.
+     * @param[out] invalidCerts invalid certificate URLs.
+     * @param[out] invalidKeys URLs of invalid certificate keys.
+     * @return Error.
+     **/
+    virtual Error ValidateCertificates(Array<CertInfo>& validInfo, Array<StaticString<cURLLen>>& invalidCerts,
+        Array<StaticString<cURLLen>>& invalidKeys)
+        = 0;
+
+    /**
+     * Destroys object instace.
+     */
+    virtual ~SecureStorageItf() = default;
+};
+
+/**
+ * Provides API to manage certificates of the same IAM certificate type.
+ */
+class CertModule {
+public:
+    /**
+     * Listener to be notified whenever certificate is added/removed from the module
+     */
+    class CertListener {
+    public:
+        /**
+         * Called whenever certificate is added to the module.
+         *
+         * @param certType certificate type.
+         * @param certInfo certificate info.
+         * @return void.
+         */
+        virtual void onCertAdded(const String& certType, const CertInfo& certInfo) = 0;
+
+        /**
+         * Called whenever certificate is removed from the module.
+         *
+         * @param certType certificate type.
+         * @param certURL certificate URL.
+         * @return void.
+         */
+        virtual void onCertRemoved(const String& certType, const String& certURL) = 0;
+
+        /**
+         * Destroys object instance.
+         */
+        virtual ~CertListener() = default;
+    };
+
+    /**
+     * Creates a new object instance.
+     *
+     * @param csrProvider a reference to x509 csr provider.
+     * @param certProvider a reference to x509 certificate provider.
+     * @param secureStorage a reference to secure storage.
+     */
+    CertModule(crypto::x509::CSRProviderItf& csrProvider, crypto::x509::CertificateProviderItf& certProvider,
```

Comment:
All interfaces should passed in `init` method.  Also, all commits should be buildable and should not break the current build and functionality: https://kb.epam.com/display/EPMDAEPRA/Coding+style+and+commits+submission